### ISSUE TITLE
fish will be stuck when tring to TAB to complete the path in scp command

### ISSUE THREAD

**Issue by c02y** [fish-shell/fish-shell#4511]
- fish, version 2.6.0
- CentOS 6.5 32-bit kernel 2.6.32-431.el6.i686
- $TERM xterm-256color

For example, when you are typing `scp ~/Pu` (with or without sudo) and hit `TAB` to complete the `Public` or anything in the path, fish will be be stuck for seconds.

This will happen even when `sh -c 'env HOME=$(mktemp -d) fish'` (without third-party customizations). But tab to complete works smoothly in bash-shell.

**Comment by ridiculousfish** [comment]
I hope Control-C recovers?

**Comment by c02y** [comment]
Control-c will cancel the current command string in the shell prompt, but cancel it only when I hit it the fourth time.

**Comment by zanchey** [comment]
Duplicate of #23, I think - although completing local files shouldn't be slow.

**Comment by faho** [comment]
Note that this is most likely due to `__fish_print_hostnames` being slow:

```fish
complete -c scp -d Hostname -n "commandline --cut-at-cursor --current-token | string match -v '*:*'" -a "
(__fish_print_hostnames):

(
	#Prepend any username specified in the completion to the hostname
	commandline -ct |sed -ne 's/\(.*@\).*/\1/p'
)(__fish_print_hostnames):
"
```

It tries to complete hostnames if there is no ":" in the token, by calling `__fish_print_hostnames` twice. Now, we have sped that up significantly via #4378 and 3b3bcc998e9c27b910f362eb9b6f24f752b4ca5c, but calling it twice (and calling `sed`) is still unnecessary.

@c02y: You'll find that 2.7b1 is faster here, but I think we can do better still.

**Comment by faho** [comment]
@c02y: 60248e8 should improve this a bit further.

The biggest possible improvement I can see would be if we could stop completing hostnames altogether in your case (e.g. via seeing that the token can't be a valid hostname), but we'll leave that for later.

**Comment by c02y** [comment]
I'm currently using v2.7b1(the 2.7b1 branch), it is the same result, even without completing hostname like this(if I understood the "stop completing hostnames" correctly):

    scp filen(TAB)

I tried to use tab to complete the files starting with filen, even there is only one file such as filename in the current directory, it will be interrupted for seconds.

-----------------------------------------

UPDATE:

I tried to compile and install the latest code from git master branch which contains the 60248e8 commit, but there were some errors from configuration files since the version of `fish` is `2.6.0-750-ged0a57ff`, so I executed fish using `sh -c 'env HOME=$(mktemp -d) fish'`, `scp` was still be interrupted for seconds, and there were errors after the interruption was gone:

in fish-shell git directory:

    scp fi(TAB to complete "fish")

after seconds I got this:

    scp ficd: The directory “/tmp/tmp.mOqbj3xFQP/.ssh” does not exist
    Standard input (line 21):
                builtin cd $relative_path
                ^
    in function “_recursive”
            called on standard input
    
    in function “_ssh_include”
            called on line 2 of file /usr/share/fish/functions/__fish_print_hostnames.fish
            with parameter list “/tmp/tmp.mOqbj3xFQP/.ssh/config”
    
    in command substitution
            called on line 0 of file /usr/share/fish/functions/__fish_print_hostnames.fish
    
    in function “__fish_print_hostnames”
            called on line 2 of file /usr/share/fish/functions/__fish_complete_user_at_hosts.fish
    
    in command substitution
            called on line 0 of file /usr/share/fish/functions/__fish_complete_user_at_hosts.fish
    
    in function “__fish_complete_user_at_hosts”
            called on standard input
    
    in command substitution
            called on standard input

**Comment by faho** [comment]
>it is the same result

I would expect it to be a bit faster, at least. Not as fast as just completing local files, because it doesn't, but faster than 2.6.0.

Now, I don't believe there's a great solution to make figuring out hostnames faster, but we should be able to figure out that something can't be a hostname.

There's https://stackoverflow.com/questions/106179/regular-expression-to-match-dns-hostname-or-ip-address#106223, which provides the following regex:

```regex
ValidHostnameRegex = "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$";
```

However, that doesn't include IPv6 addresses (which here would also be completed via known_hosts, so they'd count as hostnames for this purpose), and IMHO looks overly complicated.

So I'd propose something like `^[a-zA-Z0-9][a-zA-Z0-9._:-]*$` (i.e. start with an alphanumeric character and then consist of alphanumeric characters, dots, underlines, colons and dashes otherwise). We would then only attempt host completion if the current token matches that. (Well, username@that)

>`scp ficd: The directory “/tmp/tmp.mOqbj3xFQP/.ssh” does not exist`

Yeah, `__fish_print_hostnames` uses `cd` internally but doesn't check if the dir actually exists.

**Comment by faho** [comment]
As it turns out, these completions hit some limitations in `string`, especially with large host lists. #4610 (merged as 2de38ef8bfd) should improve this by a factor of 3 or so. I can't see any other low-hanging fruit here (of course better performance is always nice), so I'm closing this.

**Comment by c02y** [comment]
Yes, it works smoothly like other commands such as `ls`. Thank you.
